sql/sql_select.cc

```diff
@@ -27147,15 +27147,48 @@ static bool setup_sum_funcs(THD *thd, Item_sum **func_ptr)
 }
 
 
-static bool prepare_sum_aggregators(THD *thd,Item_sum **func_ptr,
+bool JOIN::prepare_sum_aggregators(THD *thd,Item_sum **func_ptr,
                                     bool need_distinct)
 {
   Item_sum *func;
   DBUG_ENTER("prepare_sum_aggregators");
   while ((func= *(func_ptr++)))
   {
+    bool need_distinct_aggregator= need_distinct && func->has_with_distinct();
+    if (need_distinct_aggregator && table_count - const_tables == 1)
+    {
+      /*
+        Check if we can optimize aggregation and avoid de-duplication
+        of values. If there is:
+        - only one table involved in the join
+        - all arguments of the aggregate function are fields (not functions)
+        - there is a unique index on all those fields,
+        then the values retrieved are guaranteed to be unique, so there's
+        no need to do the de-duplication and we can employ SIMPLE_AGGREGATOR
+      */
+      List<Item> fields_list;
+      bool all_args_are_fields= true;
+      for (uint i= 0; i < func->argument_count(); i++)
+      {
+        if (func->arguments()[i]->type() == Item::FIELD_ITEM)
+        {
+          fields_list.push_back(func->arguments()[i]);
+        }
+        else
+        {
+          all_args_are_fields= false;
+          break;
+        }
+      }
+
+      if (all_args_are_fields &&
+          list_contains_unique_index(join_tab[const_tables].table,
+                                     find_field_in_item_list,
+                                     (void *) &fields_list))
+        need_distinct_aggregator= false;
```

Comment:
Enhancement to handle case:

`SELECT a,COUNT(DISTINCT b) FROM t3 GROUP BY a;`
```
-      if (all_args_are_fields &&
-          list_contains_unique_index(join_tab[const_tables].table,
+      if (all_args_are_fields)
+      {
+         for (ORDER *tmp_group= group_list; tmp_group ; tmp_group= tmp_group->next)
+            if ((*tmp_group->item)->type() == Item::FIELD_ITEM)
+              fields_list.push_back(*tmp_group->item);
+         if (list_contains_unique_index(join_tab[const_tables].table,
                                      find_field_in_item_list,
                                      (void *) &fields_list))
-        need_distinct_aggregator= false;
+            need_distinct_aggregator= false;
+      }
```

Reply:
is there an `else` condition here that can `break` out of the `func`  list if any of the functions don't support a SIMPLE_AGGREGATOR?

Reply:
The aggregator is set separately for each function, no need to `break` the loop

Reply:
Good point with `GROUP BY`! Added it to the next revision